networkx/algorithms/bipartite/matching.py

```diff
@@ -582,3 +585,310 @@ def minimum_weight_full_matching(G, top_nodes=None, weight="weight"):
+    {3, 4}
+
+    >>> _neighbours_of_set(G, {4})
+    {0, 1, 2}
+    """
+    ret_set = {}
+    for node in node_set:
+        ret_set.update(G[node])
+
+    return set(ret_set)
+
+
+def _M_alternating_sequence(G, M, *, top_nodes=None):
+    r"""Generates a M-alternating-sequence of a bipartite graph `G` with regard to a matching M.
+
+    Two sets are calculated with the following recursive definition:
+
+    X_0 = nodes in G unmatched by M.
+    for i>=1:
+        Latex:
+            Y_i = N_{G\M}(X_{i-1})
+            X_i = N_M(Y_i)
+        English:
+            Y_i -> neighbours of X_{i-1} that were not matched by M group-subtracting all the previous Y_k's we've seen
+            X_i -> neighbours of Y_i that were matched by M
+
+    The iterative pattern stops when either Y_k or X_k are an empty set for a certain k, forever staying an empty set from k onward
+    the alternating sequece is proven to always stop in the article.
+
+    >>> G = nx.complete_bipartite_graph(3,3)
+    >>> M = nx.bipartite.hopcroft_karp_matching(G)
+    >>> _M_alternating_sequence(G, M)
+    ((), ())
+
+    >>> G = nx.Graph([(0,3),(3,0),(0,4),(4,0),(1,4),(4,1),(2,4),(4,2)])
+    >>> M = {0:3, 3:0, 1:4, 4:1}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}), ({4},))
+
+    >>> G = nx.Graph([(0,3),(3,0),(1,3),(3,1),(1,4),(4,1),(2,4),(4,2)])
+    >>> M = {0:3,3:0,4:1,1:4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph([(0,3),(3,0),(1,3),(3,1),(1,4),(4,1),(2,4),(4,2)])
+    >>> M = {0:3,3:0,4:1,1:4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph([(0,6),(6,0),(1,6),(6,1),(1,7),(7,1),(2,6),(6,2),(2,8),(8,2),(3,9),(9,3),(3,6),(6,3),(4,8),(8,4),(4,7),(7,4),(5,9),(9,5)])
+    >>> M = {0:6,6:0,1:7,7:1,2:8,8:2,3:9,9:3}
+    >>> _M_alternating_sequence(G, M)
+    (({4, 5}, {1, 2, 3}, {0}), ({8, 9, 7}, {6}))
+
+    """
+    X, Y = nx.bipartite.sets(G, top_nodes=top_nodes)
+    X_0 = set(X) - set(M.keys())
+    if X_0 == set():
+        return (), ()
+
+    X_subsets = [X_0]
+    Y_subsets = []
+
+    Y_subgroups_accumulate = set()
+    index = 1
+
+    while len(X_subsets) == 1 or X_subsets[-1] != set() and Y_subsets[-1] != set():
+        GMinusM = nx.subgraph_view(
+            G,
+            filter_edge=lambda u, v: (u, v) not in M.items()
+            and (v, u) not in M.items(),
+        )
+        Y_current = (
+            _neighbours_of_set(GMinusM, X_subsets[-1])
+        ) - Y_subgroups_accumulate
+        if Y_current == set():
+            break
+        G_M = nx.subgraph_view(
+            G, filter_edge=lambda u, v: (u, v) in M.items() and (v, u) in M.items()
+        )
+        X_current = _neighbours_of_set(G_M, Y_current)
+        if X_current == set():
+            break
+
+        Y_subsets.append(Y_current)
+        X_subsets.append(X_current)
+
+        Y_subgroups_accumulate.update(Y_current)
+        index += 1
```

Comment:
`index` doesn't seem to be used anywhere.

Reply:
You're right, it seems to serve no purpose, This is probably an artifact from the past :P

Reply:
I added a commit that removes this index variable.